library/alloc/src/vec/is_zero.rs

```diff
@@ -43,25 +46,45 @@ impl_is_zero!(f64, |x: f64| x.to_bits() == 0);
 // `IsZero` cannot be soundly implemented for pointers because of provenance
 // (see #135338).
 
+unsafe impl<T, const N: usize> IsZero for [T; N] {
+    #[inline]
+    default fn is_zero(&self) -> bool {
+        // If the array is of length zero,
+        // then it doesn't actually contain any `T`s,
+        // so `T::clone` doesn't need to be called,
+        // and we can "zero-initialize" all zero bytes of the array.
+        N == 0
+    }
+}
+
 unsafe impl<T: IsZero, const N: usize> IsZero for [T; N] {
     #[inline]
     fn is_zero(&self) -> bool {
-        // Because this is generated as a runtime check, it's not obvious that
-        // it's worth doing if the array is really long. The threshold here
-        // is largely arbitrary, but was picked because as of 2022-07-01 LLVM
-        // fails to const-fold the check in `vec![[1; 32]; n]`
-        // See https://github.com/rust-lang/rust/pull/97581#issuecomment-1166628022
-        // Feel free to tweak if you have better evidence.
-
-        N <= 16 && self.iter().all(IsZero::is_zero)
+        if T::IS_ZST {
+            // If T is a ZST, then there is at most one possible value of `T`,
+            // so we only need to check one element for zeroness.
+            // We could probably just return `true` here, since implementing
+            // `IsZero` for a zero-sized type such that `self.is_zero()` returns
+            // `false` would be useless, but to be safe we check anyway.
```

Comment:
I thought so too at first, but that would conflict with the above implementation – any `[T; N]` now implements `IsZero`, so e.g. `[NonTrivialCloneButZST; 5]` would hit this code path but mustn't be zero-initialised.

Reply:
This is in the specialization where `T: IsZero`, so this would not be run for `T = NonTrivialCloneButZST`, right?